pkg/agent/billing.go

```diff
@@ -0,0 +1,181 @@
+		case <-pushTicker.C:
+			batch := state.drainIntoBatch(&client)
+			if err := pushBillingEvents(conf, batch); err != nil {
+				return fmt.Errorf("Error pushing billing events: %w", err)
+			}
+		case <-backgroundCtx.Done():
+			// If we're being shut down, push the latests events we have before returning.
+			batch := state.drainIntoBatch(&client)
+			if err := pushBillingEvents(conf, batch); err != nil {
+				return fmt.Errorf("Error pushing billing events: %w", err)
+			}
+			return nil
+		}
+	}
+}
+
+func (s *billingMetricsState) collect(conf *BillingConfig, store *util.WatchStore[vmapi.VirtualMachine]) {
+	now := time.Now()
+
+	old := s.present
+	s.present = make(map[billingMetricsKey]vmMetrics)
+	for _, vm := range store.Items() {
+		endpointID, ok := vm.Labels[EndpointLabel]
+		if !ok {
+			continue
+		}
+
+		key := billingMetricsKey{
+			uid:        vm.UID,
+			endpointID: endpointID,
+		}
+		presentMetrics := vmMetrics{
+			cpu: uint16(*vm.Spec.Guest.CPUs.Use),
+		}
+		if oldMetrics, ok := old[key]; ok {
+			// The VM was present from s.lastTime to now. Add a time slice to its metrics history.
+			timeSlice := metricsTimeSlice{
+				metrics: vmMetrics{
+					// strategically under-bill by assigning the minimum to the entire time slice.
+					cpu: util.Min(oldMetrics.cpu, presentMetrics.cpu),
+				},
+				// note: we know s.lastTime != nil because otherwise old would be empty.
+				startTime: *s.lastTime,
+				endTime:   now,
+			}
+
+			vmHistory, ok := s.historical[key]
+			if !ok {
+				vmHistory = vmMetricsHistory{slices: nil}
+			}
+			vmHistory.slices = append(vmHistory.slices, timeSlice)
+			s.historical[key] = vmHistory
+		}
+
+		s.present[key] = presentMetrics
+	}
+
+	s.lastTime = &now
+}
+
+func (s *billingMetricsState) drainIntoBatch(client *billing.Client) *billing.Batch {
+	batch := client.NewBatch()
+
+	for key, history := range s.historical {
+		history.forEachMergedSlice(func(slice metricsTimeSlice) {
+			batch.AddIncrementalEvent(billing.IncrementalEvent{
+				IdempotencyKey: uuid.NewString(),
+				MetricName:     MetricNameCPU,
+				Type:           "", // set on JSON marshal
+				EndpointID:     key.endpointID,
+				StartTime:      slice.startTime,
+				StopTime:       slice.endTime,
+				Value:          int(slice.metrics.cpu),
+			})
+		})
+	}
+
+	s.historical = make(map[billingMetricsKey]vmMetricsHistory)
+	return batch
+}
+
+func (h vmMetricsHistory) forEachMergedSlice(process func(metricsTimeSlice)) {
+	if len(h.slices) == 0 {
+		return
+	}
+
+	last := h.slices[0]
+	for _, slice := range h.slices[1:] {
+		if last.endTime == slice.startTime && last.metrics == slice.metrics {
+			last.endTime = slice.endTime
+		} else {
+			process(last)
+			last = slice
+		}
+
+	}
+
+	process(last)
+}
+
+func pushBillingEvents(conf *BillingConfig, batch *billing.Batch) error {
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Second*time.Duration(conf.PushTimeoutSeconds))
+	defer cancel()
+
+	return batch.Send(ctx)
```

Comment:
short timeouts are probably good enough for most cases, but I think will definitely mean that we will miss events during console deploys, and it's worth calling that out, perhaps?

Reply:
I think we can miss during console deploys even with long timeouts — this is more a retry issue imo. We can discuss above.

Reply:
Also our little flap in cloud last week, makes me realize that we'll not be able to get to the control plane in 500ms from every agent. I think we should land this, and have an issue open to fiddle with this more, before GA